File: cf/dsu/step2/a.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cctype>
#include <stack>
#include <string>
#include <map>
#include <queue>
#include <set>

using namespace std;

int parents[1000005];
int mx[1000005];

void init(int n){
    for(int i = 1; i <= n+1; i++){
        parents[i] = i;
        mx[i] = i;
    }
}

int find(int a){
    if(a == -1){
        return -1;
    }
    if(parents[a] != a){
        parents[a] = find(parents[a]);
    }
    mx[a] = max(mx[a], mx[parents[a]]);
    return parents[a];
}

bool join(int a, int b){
    int pa = find(a);
    int pb = find(b);
    if(pa==pb){
        return false;
    }
    parents[pa] = pb;
    mx[pa] = max(mx[pa], mx[parents[pa]]);
    return true;
}

int main(){
    int n,m;
    cin >> n >> m;
    init(n);
    while(m--){
        char op;
        int person;
        cin >> op >> person;
        if(op == '?'){
            int cand = mx[person];
            if(cand != n+1){
                cout << cand << endl;
            }else{
                cout << -1 << endl;
            }
        }else if(op == '-'){
            join(person, person+1);
            mx[person] = max(mx[person], mx[parents[person]]);
        }
    }
}

File: cf/dsu/step2/a.py
class UnionFind:

    def __init__(self):
        self.parents = {}
        self.sz = {}
        self.size = 0
        self.mx = {}


    def add(self, a):
        if a in self.parents:
            return
        self.size += 1
        self.sz[a] = 1
        self.parents[a] = a 
        self.mx[a] = a

    def find(self, a):
        if a == -1:
            return -1
        if self.parents[a] != a:
            self.parents[a] = self.find(self.parents[a])

        return self.parents[a]

    def join(self, a, b):
        pa, pb = self.find(a), self.find(b)
        if pa == pb:
            return False
        if self.sz[pa] > self.sz[pb]:
            self.parents[pb] = pa
            self.sz[pa] += self.sz[pb]
            self.mx[pa] = max(self.mx[pa], self.mx[pb])
        else:
            self.parents[pa] = pb 
            self.sz[pb] += self.sz[pa]
            self.mx[pb] = max(self.mx[pa], self.mx[pb])
        self.size -= 1

        return True



n,m = (map(int,input().split()))
uf = UnionFind()
for i in range(1,n+2):
    uf.add(i)

for i in range(m):
    query = input().split()
    op, person = query[0], int(query[1])
    if op == "?":
        cand = uf.mx[uf.find(person)]
        if cand != n+1:
            print(cand)
        else:
            print(-1)
    elif op == "-":
        uf.join(person, person + 1)
